nixos/xen: declarative `libxenlight`, part I

SigmaSquadron opened the pull request:
- Splits up the Xen module into smaller files.
- Makes the `oxenstored` configuration compliant with RFC 0042.
- Creates the `xl` format generator.
- Makes `/etc/xl.conf` configurable.

No guest stuff yet. That's part II.

I would like to apologise in advance to whoever reviews `formats.nix`.
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

SigmaSquadron committed:
formats.xenLight: init

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
nixos/xen: init Domain 0 xl configuration module

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
nixos/xen: move Xen-related files to the xen/ subdirectory

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
nixos/xen: Separate the Xen Store options from the main Xen module

Also turns the oxenstored configuration compliant with RFC 0042.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron committed:
nixos/xen: Separate the EFI module from the main Xen module

Also adds a warning for Lanzaboote users.

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

hehongbo commented on nixos/modules/virtualisation/xen/xl.nix:
```diff
@@ -0,0 +1,55 @@
+  config = mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = cfg.dom0Resources.memory != 0 -> !cfg.settings.autoballoon == "on";
```
The type and the default value of this `autoballoon` is not defined, so if I have `dom0Resources` I will receive `error: attribute 'autoballoon' missing`, and if I set `autoballoon = "off";` I will get `error: expected a Boolean but found a string: "off"`

SigmaSquadron committed:
nixos/xen: simplify main module

- Deduplicates the removed options functions
- Inherits everything from the top-level `lib`.
- Clarifies that the maintainers are the xen team members, not just `members`.
- Removes some superfluous newlines

Signed-off-by: Fernando Rodrigues <alpha@sigmasquadron.net>

SigmaSquadron commented:
Commits have been fixed. It should now be much easier to review commit by commit.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2144

Stunkymonkey commented on pkgs/pkgs-lib/formats.nix:
if you want to add something to `pkgs/pkgs-lib/*` please provide docs at `nixos/doc/manual/development/settings-options.section.md`

SigmaSquadron replied:
Done! There's now documentation for all of the quirks of this function.

SigmaSquadron commented:
Something is wrong with the CI, and I'm not entirely sure why. The manual fails because it tries to connect to the internet for some reason, and I got emails about the basic Nix parsing checks failing, even though this is valid Nix.

SigmaSquadron commented:
@Stunkymonkey: may I ask you for a re-review?

last-genius commented on nixos/modules/virtualisation/xen/oxenstored.nix:
```diff
@@ -0,0 +1,142 @@
+# Xen Project Hypervisor OCaml-based Xen Store Daemon configuration.
+# Used in dom0 by default, but can also be used in a stubdomain.
```
This comment slightly confused me.

Only C-xenstored can run as a stubdomain at the moment, hence this language in `sysconfig/xencommons`:
```
## Type: string
## Default: xenstored
#
# Select xenstore implementation, this can be either
# of these below.
# Only evaluated if XENSTORETYPE is "daemon". <------ (i.e. if "domain", then only cxenstored)
#
# This can be either of:
#  * /usr/sbin/oxenstored
#  * /usr/sbin/xenstored
```

I don't think the current config allows choosing XENSTORETYPE either, unless I'm missing something (I hope you'll excuse me, I'm coming at this from the xen side, and not terribly familiar with nix. This is a great and much appreciated effort, eventually being able to declaratively declare domains in nix would be great!)

SigmaSquadron replied:
Hey, thanks for reviewing! I'm a bit busy with other projects at the moment, and I haven't gotten around to fixing the Xen Store config yet.

You're right, this is probably very wrong as it stands, and so far I don't think anyone is running the store in a NixOS stubdom. When I have more time again, I'll look into this PR again.

jackrosenberg commented on pkgs/pkgs-lib/formats.nix:
```diff
@@ -988,4 +988,171 @@ optionalAttrs allowAliases aliases
+                    removePrefix "{" (removeSuffix "}" v)
+                  else
+                    ''"${v}"''
+                else if v == true then
+                  1
+                else if v == false then
+                  0
```
```suggestion
                else if v then
                  1
                else if !v then
                  0
```
or will this break?
